go/vt/vttablet/tabletserver/throttle/throttler.go

```diff
@@ -0,0 +1,687 @@
+					// frequent
+					if !throttler.isDormant() {
+						throttler.collectMySQLMetrics(ctx)
+					}
+				}
+			}
+		case <-mysqlDormantCollectTicker.C:
+			{
+				if atomic.LoadInt64(&throttler.isLeader) > 0 {
+					// infrequent
+					if throttler.isDormant() {
+						throttler.collectMySQLMetrics(ctx)
+					}
+				}
+			}
+		case metric := <-throttler.mysqlThrottleMetricChan:
+			{
+				// incoming MySQL metric, frequent, as result of collectMySQLMetrics()
+				throttler.mysqlInventory.InstanceKeyMetrics[metric.GetClusterInstanceKey()] = metric
+			}
+		case <-mysqlRefreshTicker.C:
+			{
+				// sparse
+				if atomic.LoadInt64(&throttler.isLeader) > 0 {
+					go throttler.refreshMySQLInventory(ctx)
+				}
+			}
+		case probes := <-throttler.mysqlClusterProbesChan:
+			{
+				// incoming structural update, sparse, as result of refreshMySQLInventory()
+				throttler.updateMySQLClusterProbes(ctx, probes)
+			}
+		case <-mysqlAggregateTicker.C:
+			{
+				if atomic.LoadInt64(&throttler.isLeader) > 0 {
+					throttler.aggregateMySQLMetrics(ctx)
+				}
+			}
+		case <-throttledAppsTicker.C:
+			{
+				if atomic.LoadInt64(&throttler.isLeader) > 0 {
+					go throttler.expireThrottledApps()
+				}
+			}
+		}
+	}
+}
+
+func (throttler *Throttler) collectMySQLMetrics(ctx context.Context) error {
+	// synchronously, get lists of probes
+	for clusterName, probes := range throttler.mysqlInventory.ClustersProbes {
+		clusterName := clusterName
+		probes := probes
+		go func() {
+			// probes is known not to change. It can be *replaced*, but not changed.
+			// so it's safe to iterate it
+			for _, probe := range *probes {
+				probe := probe
+				go func() {
+					// Avoid querying the same server twice at the same time. If previous read is still there,
+					// we avoid re-reading it.
+					if !atomic.CompareAndSwapInt64(&probe.QueryInProgress, 0, 1) {
+						return
+					}
+					defer atomic.StoreInt64(&probe.QueryInProgress, 0)
+					throttleMetrics := mysql.ReadThrottleMetric(probe, clusterName)
+					throttler.mysqlThrottleMetricChan <- throttleMetrics
+				}()
+			}
+		}()
+	}
+	return nil
+}
+
+// refreshMySQLInventory will re-structure the inventory based on reading config settings, and potentially
+// re-querying dynamic data such as HAProxy list of hosts
+func (throttler *Throttler) refreshMySQLInventory(ctx context.Context) error {
+	log.Infof("refreshing MySQL inventory")
+
+	addInstanceKey := func(key *mysql.InstanceKey, clusterName string, clusterSettings *config.MySQLClusterConfigurationSettings, probes *mysql.Probes) {
+		for _, ignore := range clusterSettings.IgnoreHosts {
+			if strings.Contains(key.StringCode(), ignore) {
+				log.Infof("Throttler: instance key ignored: %+v", key)
+				return
+			}
+		}
+		if !key.IsValid() {
+			log.Infof("Throttler: read invalid instance key: [%+v] for cluster %+v", key, clusterName)
+			return
+		}
+		log.Infof("Throttler: read instance key: %+v", key)
```

Comment:
check if we need these logging on every refresh.

Reply:
good point. Looking into it.

Reply:
I feel kinda torn. If we log, it's too much; if we don't, we lose the ability to debug a customer's issues... not sure what to do.